added a tool for computing a lineage table like gi2taxonomy (but better)

bernt-matthias opened the pull request:
The python script:
- is ete3 based
- outputs either all levels or just the main levels
- if main levels are undefined in a linage then the info can be taken from sub, super, ... taxons

some points for the discussion:
- most important: both python scripts now use an sqlite data base as parameter (instead of a per user DB in ~/.etetoolkit/ which is created by ete3. When multiple users use galaxy an update can be triggered by one user which would affect the reproducibility for other users.)
- One would now needs an additional tool to create the ETE sqlite DB from the NCBI taxonomy DB. Any suggestions? An extra Data Source tool or extend the NCBI taxonomy Data Source.
- the name2taxid lookup takes only the first returned result. if names can be resolved (eg Actinobacteria) this is confusing. one could add an option that allows to use the lower/higher taxon. maybe it would be better to create an extra tool to create an name -> taxid mapping.
- "root" is excluded (which is different from gi2taxonomy) 
- fuzzy name search lookup could be enabled, but this would require a special build of pysqlite  where extensions are supported (which is not the default)

bernt-matthias committed:
added a tool for computing a linage table like gi2taxonomy (but better)
- ete3 based
- outputs either all levels or just the main levels
- if main levels are undefined in a linage then the info can be taken from sub, super, ... taxons

bernt-matthias committed:
making travis more happy

bernt-matthias committed:
finished the lineage generator
improved the tree generator (take DB as input for reproducibility)

bernt-matthias commented:
TODOs: 
- I could not add the taxa.sqlite because it is to large for github. Is there a solution? 
- Currentlt the test for the tree generator fails and I do not understand why. 

What do you think: @nsoranzo @anilthanki

bernt-matthias committed:
make travis happier

bernt-matthias commented:
Note the ete version where a sqlite DB creation from the NCBI dump is possible is apparently not available via pip: https://github.com/etetoolkit/ete/issues/282

nsoranzo commented on tools/ete/ete_lineage_generator.py:
```diff
@@ -0,0 +1,129 @@
+    for l in lineage:
+        if not lineage_ranks[l] in RANK_IDX:
+            continue
+        if ranks[ RANK_IDX[lineage_ranks[l]]] != "NA":
```
Can you remove the space between `[` and `RANK_IDX` to make flake8 happy, please?

bernt-matthias committed:
removed space between [ and RANK_IDX to make flake8 happy

nsoranzo commented:
> I could not add the taxa.sqlite because it is to large for github. Is there a solution?

@bernt-matthias Can you reduce the size of the database by deleting the table/rows that are not used in the tests?

bernt-matthias committed:
added small taxdump test data
temporarily switched to local ete library

nsoranzo commented on tools/ete/ete_init_taxdb.py:
```diff
@@ -0,0 +1,7 @@
+import sys
+sys.path.append("./ete/")
```
Why is that needed?

bernt-matthias replied:
see my comment below (the official ete code does not work like this)

nsoranzo commented on tools/ete/ete_init_taxdb.py:
```diff
@@ -0,0 +1,7 @@
+import ete3
+
+ete3.NCBITaxa(taxdump_file="/tmp/taxdump/taxdump.tar.gz", dbfile="taxdump.sqlite")
+
```
Remove blank line for flake8.

bernt-matthias replied:
done

bernt-matthias commented:
The testdata contains only the lineages of the input data.

Note that the sqlitedb can be generated with ete_init_taxdb.py only by using my fork of ete3 (https://github.com/bernt-matthias/ete PR is pending for quite a while). One could make also a hack using the normal ete3 library by calling NCBITaxa function only with the tar.gz file and then moving the sqlite file from ~/.etetoolkit ...

bernt-matthias committed:
removed local ete path

bernt-matthias committed:
removed blank line

bernt-matthias commented:
the remaining question is how we integrate the sqlite db in the tool? data manager?

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,143 @@
+<tool id="ete_linage_generator" name="ETE lineage generator" version="3.0.0b35">
```
s/linage/lineage/

bernt-matthias committed:
s/linage/lineage/

nsoranzo commented:
> the remaining question is how we integrate the sqlite db in the tool? data manager?

Do you want the user to be able to select between multiple versions of the database?

> Note that the sqlitedb can be generated with ete_init_taxdb.py only by using my fork of ete3 (https://github.com/bernt-matthias/ete PR is pending for quite a while).

Do you mean https://github.com/etetoolkit/ete/pull/288 ? I can't see why is that related. Or do you mean release 3.0.0b36 , which is not available on PyPI yet as you reported in https://github.com/etetoolkit/ete/issues/282 ?

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,143 @@
+<tool id="ete_lineage_generator" name="ETE lineage generator" version="3.0.0b35">
+		<description>from a list of species/taxids using the ETE Toolkit</description>
```
Indentation. Please always use 4-spaces instead of tabs.

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,143 @@
+<tool id="ete_lineage_generator" name="ETE lineage generator" version="3.0.0b35">
+		<description>from a list of species/taxids using the ETE Toolkit</description>
+     <requirements>
```
Use 4-spaces instead of 5

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,143 @@
+        <!-- Anything other than zero is an error -->
+        <exit_code range="1:" />
+    </stdio>
+    <command>
```
You can add `detect_errors="exit_code"` attribute to `<command>` and remove the whole `<stdio>` block.

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,143 @@
+-d '$database'
+-o '$outputFile'
+#if $ranks.levels == "full"
+-f 
```
Can you add (4-spaces) indentation also to the Cheetah code, please?

bernt-matthias commented:
> Do you want the user to be able to select between multiple versions of the database?

Yes, I think that this would be a good idea. (In the current tool the sqlite DB in ~/.etetoolkit is used which is particularly problematic for when jobs are started as real user)

bernt-matthias commented:
> Do you mean etetoolkit/ete#288 ? I can't see why is that related. Or do you mean release 3.0.0b36 , which is not available on PyPI yet as you reported in etetoolkit/ete#282 ?

I think I mixed up things. Indeed I mean https://github.com/etetoolkit/ete/issues/282

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,143 @@
+            </when>
+	</conditional>
+		<conditional name="compresscond">
+				<param name="compress" type="boolean" label="Fill unnamed ranks" truevalue="-c" falsevalue="" help="Fill unnamed ranks with super/sub ranks (see -l)" />
```
Please use a `type="select"` for the conditional param, see https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#booleans .

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,143 @@
+		</conditional>
+    </inputs>
+    <outputs>
+        <data format="tsv" name="outputFile" label="${tool.name} on ${on_string}"/>
```
Move `format` after `name`.

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,143 @@
+		<conditional name="compresscond">
+				<param name="compress" type="boolean" label="Fill unnamed ranks" truevalue="-c" falsevalue="" help="Fill unnamed ranks with super/sub ranks (see -l)" />
+				<when value="-c">
+					<param name="lower" type="boolean" label="Prefer lower ranks for filling" truevalue="-l" falsevalue="" help="Prefer lower levels for filling" />
```
`label` and `help` should always be the second-to-last and last attributes respectively, according to IUC [coding style](https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#coding-style). Rationale is that if the XML element is long, `label` and `help` attributes can be on a new line(s).

bernt-matthias committed:
coding style changes

nsoranzo commented:
> > Do you want the user to be able to select between multiple versions of the database?

> Yes, I think that this would be a good idea. (In the current tool the sqlite DB in ~/.etetoolkit is used which is particularly problematic for when jobs are started as real user)

So a good start would be to add a data table (adding a `tool_data_table_conf.xml.sample` and a .`loc` file in `tool-data/` ) to be used for the `database` param with the `<options from_data_table="...">` syntax.

See https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/integration_checklist.html#creating-the-tool-wrapper

bernt-matthias committed:
added ete_init_taxdb

bernt-matthias commented:
I decided against the tool_data_table_conf option. One can still get the data base as library. 

I added a tool for creating the sqlite DB. But this wont work until https://github.com/etetoolkit/ete/issues/282 is fixed. 

There are currently lots of errors popping up in planemo test and serve: 
```
2017-09-12 11:21:30,000 WARNI [galaxy.tools] Tool ete_lineage_generator: a when tag has not been defined for 'ranks (levels) --> full', assuming empty inputs.
2017-09-12 11:21:30,000 WARNI [galaxy.tools] Tool ete_lineage_generator: a when tag has not been defined for 'ranks (levels) --> primary', assuming empty inputs.
2017-09-12 11:21:30,001 WARNI [galaxy.tools] Tool ete_lineage_generator: a when tag has been defined for 'compresscond (compress) --> -c', but does not appear to be selectable.
```

`ValueError: ('No case matched value:', 'compresscond', 'None')
`

Any suggestion what I did wrong?

nsoranzo commented on tools/ete/ete_init_taxdb.xml:
```diff
@@ -0,0 +1,52 @@
+            </param>
+            <when value="local">
+                <param name="taxdump" type="data" format="txt" label="NCBI taxdump" help="NCBI taxonomy database dump in tar.gz format" />
+            </when>
```
Add:
```xml
            <when value="latest" />
```

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,144 @@
+                <option value="primary" selected="true">Primary</option>
+                <option value="full">Full</option>
+                <option value="manual">Manual</option>
+            </param>
```
Add:
```xml
            <when value="primary" />
            <when value="full" />
```

bernt-matthias commented:
And one more question: how to handle the tar.gz file correctly? Is "text" OK?

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,144 @@
+            </when>
+        </conditional>
+        <conditional name="compresscond">
+                <param name="compress" type="select" truevalue="-c" falsevalue="" label="Fill unnamed ranks" help="Fill unnamed ranks with super/sub ranks (see -l)" />
```
That's an hybrid now ;)
It should be:
```xml
            <param name="compress" type="select" label="Fill unnamed ranks" help="Fill unnamed ranks with super/sub ranks (see -l)">
                <option value="-c">Yes</option>
                <option value="">No</option>
            </param>
```

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,144 @@
+                        <option value="-l">Yes</option>
+                        <option value="">No</option>
+                    </param>
+                </when>
```
Add:
```xml
                <when value="" />
```

nsoranzo commented on tools/ete/ete_species_tree_generator.xml:
```diff
@@ -11,17 +11,19 @@
     </command>
     <inputs>
         <param name="speciesFile" type="data" format="txt" label="Species file" help="List with one species per line" />        
+		<param name="database" type="data" format="sqlite" label="(ETE3) Taxonomy Database" help="The sqlite formatted Taxonomy used by ETE3 (which is derived from NCBI taxonomy)" />
```
Don't use tabs, please!

bgruening commented:
You could use: https://github.com/galaxyproject/galaxy/blob/dev/config/datatypes_conf.xml.sample#L200

bernt-matthias committed:
added missing whens and removed tab

nsoranzo commented:
> I decided against the tool_data_table_conf option. One can still get the data base as library.
> I added a tool for creating the sqlite DB.

That's fine for me, I actually wanted to suggest you something like that, but forgot yesterday.

I had review comments that should fix the errors you're seeing.

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,147 @@
+            <when value="full" />
+        </conditional>
+        <conditional name="compresscond">
+                <param name="compress" type="select" label="Fill unnamed ranks" help="Fill unnamed ranks with super/sub ranks (see -l)" />
```
Still not right, this line should be substituted with:
```xml
            <param name="compress" type="select" label="Fill unnamed ranks" help="Fill unnamed ranks with super/sub ranks (see -l)">
                <option value="-c">Yes</option>
                <option value="">No</option>
            </param>
```

Also the indentation inside this conditional block is wrong.

bernt-matthias committed:
added missing options

bernt-matthias committed:
fixed indentation

nsoranzo commented on tools/ete/ete_init_taxdb.py:
```diff
@@ -0,0 +1,26 @@
+if options.database is None:
+    parser.error("-d option must be specified")
+if options.taxdump is not None:
+    taxdump = options.database
```
`options.taxdump` should go here, I suppose.

nsoranzo commented on tools/ete/ete_init_taxdb.xml:
```diff
@@ -0,0 +1,53 @@
+    <tests>
+        <test>
+            <param name="database" ftype="sqlite" value="taxdump.sqlite" />
+            <output name="taxdump" file="taxdump.tar.gz" />
```
You seem to have swapped the param and output here.

bernt-matthias committed:
swapped param and output, swapped taxdump and database

nsoranzo commented on tools/ete/ete_init_taxdb.xml:
```diff
@@ -0,0 +1,53 @@
+    </outputs>
+    <tests>
+        <test>
+            <param name="taxdump" ftype="tar" file="taxdump.tar.gz" />
```
s/file/value/

nsoranzo commented on tools/ete/ete_init_taxdb.xml:
```diff
@@ -0,0 +1,53 @@
+    <tests>
+        <test>
+            <param name="taxdump" ftype="tar" file="taxdump.tar.gz" />
+            <output name="database" value="taxdump.sqlite" />
```
s/value/file/

bernt-matthias committed:
file <-> value

nsoranzo commented:
Tests are failing because Planemo cannot validate 2 URLs, will be fixed by https://github.com/galaxyproject/planemo/pull/725 .

bernt-matthias committed:
implemented a taxdump update that works

- also foor the currently available version of ete3
- improved (but still dysfunctional tests)

bernt-matthias commented:
I guess its nearly done. Still have some problems with the tests.

nsoranzo commented on tools/ete/ete_init_taxdb.py:
```diff
@@ -0,0 +1,26 @@
+import optparse
+import sys
```
This import is unused, please remove it (or comment it out as the next line).

nsoranzo commented on tools/ete/ete_init_taxdb.py:
```diff
@@ -0,0 +1,26 @@
+    urlretrieve("http://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz", "taxdump.tar.gz")
+    taxdump = "taxdump.tar.gz"
+
+# will remove a taxdump.tar.gz file at the end 
```
Remove trailing whitespace.

nsoranzo commented on tools/ete/ete_init_taxdb.py:
```diff
@@ -0,0 +1,26 @@
+
+# will remove a taxdump.tar.gz file at the end 
+# which will lead to an errmessage if not present
+# if the tool is run on a taxdump in the current dir it will be 
```
Remove trailing whitespace.

nsoranzo reviewed (commented):
3 linting fixes needed, then we'll see the proper tool tests

bernt-matthias committed:
removed trailing whitespaces

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,153 @@
+        </test>
+        <test>
+            <param name="speciesFile" ftype="txt" value="species.txt" />
+            <param name="database" ftype="sqlite" value="taxsump.sqlite" />
```
s/taxsump/taxdump/

nsoranzo commented on tools/ete/ete_init_taxdb.xml:
```diff
@@ -0,0 +1,54 @@
+    </requirements>
+    <command detect_errors="exit_code">
+    <![CDATA[
+python $__tool_directory__/ete_init_taxdb.py
```
Please add single quotes around `$__tool_directory__/ete_init_taxdb.py`

nsoranzo commented on tools/ete/ete_init_taxdb.xml:
```diff
@@ -0,0 +1,54 @@
+    <![CDATA[
+python $__tool_directory__/ete_init_taxdb.py
+#if $latest == "local"
+    -t '$taxdump'
```
s/$taxdump/$latestcond.taxdump/

nsoranzo commented on tools/ete/ete_init_taxdb.py:
```diff
@@ -0,0 +1,26 @@
+if options.taxdump is not None:
+    taxdump = options.taxdump
+else:
+    from urllib import urlretrieve
```
For Python3 compatibility, can you please change this to:
```
from six.moves.urllib.request import urlretrieve
```
and move it to the top, just after the `ete3.ncbi_taxonomy` import?

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,153 @@
+    </requirements>
+    <command detect_errors="exit_code">
+    <![CDATA[
+python $__tool_directory__/ete_lineage_generator.py
```
Add single-quotes around `$__tool_directory__/ete_lineage_generator.py`

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,153 @@
+-s '$speciesFile'
+-d '$database'
+-o '$outputFile'
+#if $levels == "full"
```
s/$levels/$ranks.levels/

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,153 @@
+-o '$outputFile'
+#if $levels == "full"
+    -f 
+#elif $levels == "manual"
```
s/$levels/$ranks.levels/

nsoranzo commented on tools/ete/ete_lineage_generator.xml:
```diff
@@ -0,0 +1,153 @@
+#if $levels == "full"
+    -f 
+#elif $levels == "manual"
+    #for $rank in $manualranks:
```
s/$manualranks/$ranks.manualranks/

nsoranzo reviewed (commented):
Let me know if you need help with setting up [Planemo](https://planemo.readthedocs.io/), `planemo shed_lint` and `planemo test` can really speed up tool development!

bernt-matthias committed:
small bug fixes

- python3 compatibility
- typos
- quoting

bernt-matthias committed:
updated test data

bernt-matthias commented:
Still struggling with planemo test. Somehow there is a difference for the database and the generated tree. I never used shed_lint. Where can I start to read?

Thanks for your help.. learning slowly.

anilthanki commented on tools/ete/ete_init_taxdb.py:
```diff
@@ -0,0 +1,23 @@
+import ete3.ncbi_taxonomy
+import optparse
```
import optparse should be before ete3.ncbi_taxonomy

nsoranzo replied:
Yep, we (as `galaxy` and `tools-iuc`) use the "smarkets" style, which divides imports in 4 groups (separated by a single blank line):

- `from __future__ import ...`
- standard library imports
- third-party imports
- application-specific imports

Within each grouping, imports should be sorted lexicographically, ignoring case, according to each module's full package path, but with `import ...` statements before `from X import ...` statements, see an [example](https://github.com/PyCQA/flake8-import-order/blob/master/tests/test_cases/complete_smarkets.py).

So, in this case it should be:

```python
import optparse

import ete3.ncbi_taxonomy
from six.moves.urllib.request import urlretrieve
```

nsoranzo commented:
Planemo documentation is at http://planemo.readthedocs.io/ , the commands we use are:
```
planemo shed_lint --tools --ensure_metadata --urls --report_level warn --fail_level error --recursive TOOL_DIR
planemo test --conda_dependency_resolution --conda_auto_install --galaxy_branch release_17.05 TOOL_DIR
```

bernt-matthias committed:
fixed import order

bernt-matthias commented:
Import order is fixed. `planemo shed_lint` is successful, but `planemo test` still has problems with ete_init and ete_species_tree_generator. 

For ete_species_tree_generator the comparison of `... test-data/out.nhx v. /tmp/tmpTglW3_out.nhx` is not successful because the file in /tmp seems to be absent. The actual output seems to be correct.

nsoranzo commented:
> For ete_species_tree_generator the comparison of ... test-data/out.nhx v. /tmp/tmpTglW3_out.nhx is not successful because the file in /tmp seems to be absent. The actual output seems to be correct.

That's not your fault, the NCBI database has been updated and the tree is slightly different now. You can update the test output data with:
```
planemo test --update_test_data /PATH/TO/earlham-galaxytools/tools/ete/ete_species_tree_generator.xml
```
and then commit the changes to `tools/ete/test-data/out.nhx` . I can do that, but then you would have to rebase this PR.

bernt-matthias committed:
updated test data

bernt-matthias commented:
The problem with the species tree generator seems to be fixed. 

The init DB test still fails despite updated test data. Since the test works on a local dump of (a part of the taxonomy) a change of the online NCBI taxonomy should be not the reason. I guess that using diff on the binary sqlite files could be the problem. Are there other possibilities to test. 

It would also be nice to have a test for the download. Can one just test for existence / file size larger X?

nsoranzo commented:
@bernt-matthias Thanks for updating the `ete_species_tree_generator`

Probably the best choice is add `compare="sim_size"` to the `<output />` element in the failing test. For more details, see https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output

bernt-matthias committed:
ete_init_taxdb test based on sim_size

bernt-matthias commented:
OK - Got it. If I would like to add a test for the download of the complete taxdb how could I create a useful test without having the complete taxdump-sqlite in the test-data directory? 

The simplest test would be to test for the existence of the files. How would this be done? Alternatively there is a md5 at ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz.md5 that should match the taxdump.tar.gz file (but the taxdump is not the output, but the sqlite). Could this be integrated?

nsoranzo commented:
@bernt-matthias The possible tests are the one in the link I gave above https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output , so what I have in my mind is probably not possible at the moment.
On the other hand, tests are supposed to be fast and with small inputs/outputs, so what we have seems good enough to me!

bernt-matthias commented:
Wonderful. Thanks. 

Btw. Next week I will give a small workshop on Galaxy and wrapping galaxy tools. I will suggest there to wrap some of the ete3 command line tools...

bernt-matthias committed:
added a wrapper for ete3 mod

bernt-matthias commented:
I "finished" a wrapper for `ete mod`. There is still an error popping up due to the `--help` output in `<help>`. Could someone tell me what the problem and the solution is? I guess some quoting is necessary. Otherwise the tool seems be finished apart from a few lines of documentation in the help. 

I currently do not have the time to wrap `ete3 compare` and `ete3 view` which would be really nice.

nsoranzo commented on tools/ete/ete_mod.xml:
```diff
@@ -0,0 +1,170 @@
+        </test>
+    </tests>
+    <help><![CDATA[
+usage: ete3 mod [-h] [-t [SRC_TREES [SRC_TREES ...]]]
```
@bernt-matthias I don't think it's useful to have the tool help copied in the `<help>` section, it's better to have a general description of what the tool does and a link to the upstream documentation. Each option should be documented using the `help` attribute of the `<param>`s.

bernt-matthias committed:
fixed help and citation

bernt-matthias commented:
Agreed. Finished from my side.

nsoranzo commented on tools/ete/ete_mod.xml:
```diff
@@ -0,0 +1,125 @@
+  number={6},
+  pages={1635-1638},
+  year={2016}
+}
```
You can replace with:
```
<citation type="doi">10.1093/molbev/msw046</citation>
```

Actually it starts to make sense to have a macro file for requirements and citations.

bernt-matthias committed:
use doi citation

nsoranzo committed:
Small cleanups for ete3 tools

nsoranzo committed:
Put version, requirements and citations in macros
